### PR TITLE
Do auth in the user's browser, do auth through the dashboard, support Google Auth + others

### LINKED ISSUE #1162: Cypress test login via GitHub impossible with U2F keys
### Current behavior:

Log in errors.

### Desired behavior:

Log in works.

### How to reproduce:

Log in with a GitHub account that has U2F enabled using a hardware key.

### Additional Info (images, stack traces, etc)

image

- Operating System: Windows 10
- Cypress Version: 1.4.1
- Browser Version: Google Chrome Version 63.0.3239.132 (Official Build) (64-bit)

### LINKED ISSUE #3519: Support login via Google OAuth in Test Runner
Cypress recently introduced the ability to login to the [Cypress Dashboard via Google](https://www.cypress.io/blog/2019/02/12/new-feature-log-in-to-the-cypress-dashboard-via-google/). While this is great for allowing a wider audience to access the Dashboard, the lack of feature parity with the Test Runner creates some awkwardness ... the requirement for a GitHub account in order to setup a project to record to the Dashboard.

### LINKED ISSUE #3583: Proposal: Authenticate through user's installed browser
### Current behavior:

Logging in with GitHub opens the Oauth authorize page in Cypress's packaged Electron, which causes issues:

- #1251 - Electron goes out of date and GitHub blocks logging in with it
- #1162 - Electron doesn't support hard tokens
- users can't see the URL of the page, could potentially be a security problem
- users can't use password vaults
- #4337 - networking bugs can cause auth not to work

### Desired behavior:

Logging in with GitHub through the desktop app opens a tab in the user's default system browser where they can authorize our app

Could be built by creating a HTTP server on localhost to accept auth redirects: https://stackoverflow.com/questions/37546656/handling-oauth2-redirect-from-electron-or-other-desktop-platforms

~~Or by creating a custom protocol handler and setting the GH redirect URL to that: https://glebbahmutov.com/blog/electron-app-with-custom-protocol/~~ Won't work, registering a custom protocol globally usually requires advanced permissions.

### LINKED ISSUE #4337: Log In With GitHub Gives Blank Screen
### Current behavior:

Able to click Log In With GitHub, but a blank window appears and never loads the GitHub login. This is behind a corporate proxy, but the proxy configuration has already been set up.

LogIn

LogInBlank

### Desired behavior:

Ability to log in to GitHub to see the Dashboard Service.

### Steps to reproduce: (app code and test code)

1. Open Cypress v3.3.1
2. Click "Runs" tab.
3. Click LogIn with GitHub.
4. Get new window popup with blank screen.

### Versions

**Cypress**: _v3.3.1_
**OS**: _Windows 10_

### PR DESCRIPTION
**Pull request by flotwig**
* Fixes #3583 
* Fixes #1162
* Fixes #3519
* Fixes #4337 

### User-facing changes

* Auth no longer opens up in the Cypress test runner directly. Instead, it will open up in the user's browser by default, and if we think that failed, we just offer the user the URL directly.
* Instead of logging in directly to GitHub, the login will go through the Dashboard, which will let users log in with Google, GitHub, SSO, etc...
* "Log In with GitHub" -> "Log In to Dashboard" image

### Auth flow

1. User clicks "Log In to Dashboard"
2. Cypress retrieves auth URL (`dashboardAuthUrl`) from `https://api.cypress.io/auth`
3. Cypress spawns a local Express server on a random port
4. Cypress opens `http://127.0.0.1:${port}/redirect-to-auth` in the user's browser
	- If this URL is not hit within 5 seconds, or if it throws an exception, or if the user has tried to login without this URL being hit in the current session, this warning appears so that the user can manually open the URL: image
5. `redirect-to-auth` redirects to `${dashboardAuthUrl}?port=${port}&state=${some random string}&machineId=...&version=...&platform=...&arch=...`.
	- Once this is detected, the test runner changes so that the user knows to continue in their browser: image
6. User completes authentication on the `dashboardAuthUrl` page, and is eventually redirected to `http://127.0.0.1:${port}/auth?state=${same random string}&access_token=${access_token}&refresh_token=...&name=...&email=...`, which is a route on the local Express server
7. Cypress will parse and save this result. If it is successful, Cypress will redirect the browser to `${dashboardAuthUrl}?status=success` and re-focus Cypress. If it is not successful, Cypress will redirect the browser to `${dashboardAuthUrl}?status=error`, and re-focus Cypress, displaying the exact error inside Cypress. The Express server will also be shut down in either scenario.
	- If auth is successful it looks like this: 
image

### CODE REVIEW

**Comment by flotwig** [comment]
Here are my notes on what would be needed on the backend to make this work:

**current flow**

1. user clicks "login with GitHub"
2. TR requests login URL from `api.cypress.io/auth`
3. API sends github auth URL to TR
4. TR spawns electron window pointing to github auth URL
5. user completes github login and is redirected to cypress website with code attached
6. TR snarfs the code out of the new URL and closes the login window
7. TR makes a request to `api.cypress.io/signin` to get the token
8. API responds with token for cypress

**proposed flow**

note: this flow assumes that our auth0 server has been set up with a 2nd client separate from the dashboard client to be used with the TR with a redirect_uri set to http://localhost:62213

1. user clicks "login with github" or "login with google"
2. TR spawns express app on `localhost:62213` to listen for redirect
2. TR gets Cypress Auth0 server login URL from `api.cypress.io/auth?provider=<provider>`
3. API sends Cypress Auth0 auth URL to TR
4. TR launches browser pointing to login URL
5. user is redirected to github login, then redirected back to Cypress Auth0, then redirected to `localhost:62213`
6. TR receives `code` and focuses itself
7. TR makes a request to Cypress's token endpoint to get a token
8. API responds with a token for cypress

**questions**

- do we want to have _both_ login buttons in the TR interface? or should we send the user to Cypress's login page, which has those buttons?
    - benefit: future SSO/oauth provider changes would be easier, purely server-side changes
    - downside: 1 more click for users

**Comment by flotwig** [comment]
Taking this PR out of WIP as the changes are now available in staging and this is ready for review.

**Comment by flotwig** [comment]
@jennifer-shehane That's happening because there is a nonce sent through the auth flow that is generated when the user clicks "Log In with Dashboard", so when you have a second login opened, the first login's nonce is overwritten and no longer valid.

I don't see any security downside to just using one nonce for the entire session, so I've made it only generate one nonce at a time so this edge case can't happen.

**Comment by flotwig** [comment]
More screenshots:

- Prompt in desktop-gui: 
image
- Logging in... 
image
- Browser opened to login page: 
image
- TR and browser together: 
image
- Login success: 
image 
image
image

File: packages/server/lib/api.coffee [hunk 1]
```diff
@@ -238,31 +273,6 @@ module.exports = {
     .timeout(timeout)
     .catch(tagError)
 
-  createSignin: (code) ->
-    machineId()
```
Comment (brian-mann) [inline packages/server/lib/api.coffee]: this needs to be reimplemented in the new logic - and the API needs to ensure this property is being handled.

Comment (flotwig) [reply packages/server/lib/api.coffee]: b8e47468d0b2a56de38bec04e4bc857fbccb4628 adds it to query params that we send to the dashboard

**Pull request merged**